framework/src/utils/MooseMeshXYCuttingUtils.C

```diff
@@ -0,0 +1,1229 @@
+
+  if (improve_boundary_tri_elems)
+    boundaryTriElemImprover(mesh, new_boundary_id);
+}
+
+void
+boundaryTriElemImprover(ReplicatedMesh & mesh, const boundary_id_type boundary_to_improve)
+{
+  if (!MooseMeshUtils::hasBoundaryID(mesh, boundary_to_improve))
+    mooseError(
+        "MooseMeshXYCuttingUtils::boundaryTriElemImprover(): The boundary_to_improve provided "
+        "does not exist in the given mesh.");
+  BoundaryInfo & boundary_info = mesh.get_boundary_info();
+  auto side_list = boundary_info.build_side_list();
+  // Here we would like to collect the following information for all the TRI3 elements on the
+  // boundary: Key: node id of the off-boundary node Value: a vector of tuples, each tuple contains
+  // the following information:
+  // 1. The element id of the element that is on the boundary to improve
+  // 2. the one node id of that element that is on the boundary to improve
+  // 3. the other node id of the element that is on the boundary to improve
+  std::map<dof_id_type, std::vector<std::tuple<dof_id_type, dof_id_type, dof_id_type>>>
+      tri3_elem_info;
+  for (const auto & side : side_list)
+  {
+    if (std::get<2>(side) == boundary_to_improve)
+    {
+      Elem * elem = mesh.elem_ptr(std::get<0>(side));
+      if (elem->type() == TRI3)
+      {
+        const auto key_node_id = elem->node_id((std::get<1>(side) + 2) % 3);
+        const auto value_elem_id = elem->id();
+        const auto value_node_id_1 = elem->node_id(std::get<1>(side));
+        const auto value_node_id_2 = elem->node_id((std::get<1>(side) + 1) % 3);
+        tri3_elem_info[key_node_id].push_back(
+            std::make_tuple(value_elem_id, value_node_id_1, value_node_id_2));
+      }
+    }
+  }
+  // Elements that need to be removed
+  std::vector<dof_id_type> elems_to_remove;
+  // Now check if any group of TRI3 sharing an off-boundary node can be improved.
+  for (const auto & tri_group : tri3_elem_info)
+  {
+    // It is possible to improve only when more than one TRI3 elements share the same off-boundary
+    // node
+    std::vector<std::pair<dof_id_type, dof_id_type>> node_assm;
+    std::vector<dof_id_type> elem_id_list;
+    for (const auto & tri : tri_group.second)
+    {
+      node_assm.push_back(std::make_pair(std::get<1>(tri), std::get<2>(tri)));
+      elem_id_list.push_back(std::get<0>(tri));
+    }
+    std::vector<dof_id_type> ordered_node_list;
+    std::vector<dof_id_type> ordered_elem_list;
+    MooseMeshUtils::makeOrderedNodeList(
+        node_assm, elem_id_list, ordered_node_list, ordered_elem_list);
+    elems_to_remove.insert(
+        elems_to_remove.end(), ordered_elem_list.begin(), ordered_elem_list.end());
+
+    // For all the elements sharing the same off-boundary node, we need to know how many separated
+    // subdomains are involved
+    // If there are extra element ids defined on the mesh, they also want to retain their boundaries
+    const unsigned int n_elem_extra_ids = mesh.n_elem_integers();
```

Comment:
```suggestion
    // Only triangle elements that share a side can be merged
    const unsigned int n_elem_extra_ids = mesh.n_elem_integers();
```
(hence why no search through all block infos below)